app/models/piece.rb

```diff
@@ -3,16 +3,20 @@ class Piece < ApplicationRecord
   has_many :moves
 
   def move!(x, y)
-    return false unless valid_move?(x, y)
-
-    if castling_kingside?(x, y)
-      rook_at(7, y).update_attributes(x_position: 5, y_position: y)
-    end
-    if castling_queenside?(x, y)
-      rook_at(0, y).update_attributes(x_position: 3, y_position: y)
-    end
+    return false unless move_valid?(x, y)
+    update_rook_if_castling(x, y)
     update_attributes(x_position: x, y_position: y)
     Move.create(piece_id: id, game_id: game_id, destination_x: x_position, destination_y: y_position)
+    update_attributes(type: 'Queen') if promoting_pawn?(y)
+  end
+
+  def promoting_pawn?(y)
+    type == 'Pawn' && y == 7 || y.zero?
```

Comment:
You will need the `(` and `)` around the second grouping, I believe, as I think `&&` has higher precedence than `||`, meaning it's implying `(type == 'Pawn' && y == 7) || y.zero?`, when what you want is:

```ruby
type == 'Pawn' && (y == 7 || y.zero?)
```